### ISSUE TITLE
Generic worker actuator leaves orphaned MCM resources when deleting shoots

### ISSUE THREAD

**Issue by prashanth26** [gardener/gardener#2453]
**What happened**:
The generic worker actuator sometimes tends to leave orphan resources while shoot deletion.

**What you expected to happen**:
These resources are not expected to be left out.

**How to reproduce it (as minimally and precisely as possible)**:
Not really sure.

**Anything else we need to know**:
The creation timestamps on the orphan resources were later than shoot deletion timestamps which is weird. We will need to look into this. 

**Environment**:
Seem on Azure and other infra as well.

**Comment by prashanth26** [comment]
@timuthy - Do you have a list of resources that you observed which were orphaned? I remember you mentioning - machine classes, secrets. You didn't observe machine-deployments/machine-sets/machines right?

**Comment by timuthy** [comment]
Yes @prashanth26, here are the resoruces:

```
secrets
NAME                       TYPE     DATA   AGE
shoot--<...>-45b06-z1      Opaque   5      13h
shoot--<...>-dab76-z1      Opaque   5      13h
shoot--<...>-a2eb9-z1      Opaque   5      13h

azuremachineclasses
NAME                       VM SIZE           AGE
shoot--<...>-45b06-z1      Standard_D4s_v3   13h
shoot--<...>-dab76-z1      Standard_D2s_v3   13h
shoot--<...>-a2eb9-z1      Standard_E8s_v3   13h

managedresources
NAME                         CLASS   APPLIED   HEALTHY   AGE
extension-worker-mcm-shoot           True      True      13h
```

Timestamps:
```
ManagedResource
- creationTimestamp: "2020-06-09T16:04:45Z"
- deletionTimestamp: "2020-06-09T16:12:24Z"
MachineClasses
- creationTimestamp: "2020-06-09T16:04:45Z"
- deletionTimestamp: "2020-06-09T16:12:15Z"
Shoot
- deletionTimestamp: '2020-06-09T15:48:48Z'
- creationTimestamp: '2020-06-09T13:45:17Z'
```

**Comment by prashanth26** [comment]
I had an initial analysis of the code path on MCM at least. I see that there are two code paths that could be improved to fix such issues. 

1. Add finalizers on (provider)MachineClasses only when there is a machine/machine-set/machine-deployment refering it. Currently it adds for any machineClass without a deletion timestamp. Update [this logic](https://github.com/gardener/machine-controller-manager/blob/master/pkg/controller/awsmachineclass.go#L130-L136) in all providerMachineClass reconciliation logic.

2. Add finalizers on secrets that only have valid machines backing it. Currently, the secrets look just at the machine class object (which may not be used by any machines). Maybe we could look at even the machine object (machine class finalizers) - https://github.com/gardener/machine-controller-manager/blob/master/pkg/controller/secret.go#L67. 

These improvements should improve things. However, I still haven't analyzed how new providerMachineClasses were created after the shoot was targetted for deletion. This is probably due to the rolling update that was triggered racing with the shoot deletion. This logic still needs to be fixed and validated at the gardener/extensions.

**Comment by prashanth26** [comment]
One thought on reproducing this issue,

My guess would be to create an Azure shoot, and once creation is done. Trigger update and deletion (in parallel or in immediate successions). I think this should reproduce this.

**Comment by prashanth26** [comment]
On a similar line, 

I saw the following issue on a shoot. 

I think what has happened is that there was a work pool with valid credentials. The credentials and machine images were updated to an incorrect one. And then the credentials was again updated to the correct one. Bcoz the machine class had a deletion timestamp the credentials were not updated properly by the Gardener extension. 

NAME                                                                      READY   DESIRED   UP-TO-DATE   AVAILABLE   AGE
machinedeployment.machine.sapcloud.io/shoot--xxx-worker-z1   1       1         0            1           27h
NAME                                                                         DESIRED   CURRENT   READY   AGE
machineset.machine.sapcloud.io/shoot--xxx-worker-z1-c876ff6df   1         1         0       24h
NAME                                                                            STATUS    AGE
machine.machine.sapcloud.io/shoot--xxx-worker-z1-c876ff6df-lm4z9   Running   24h
NAME                                                                          MACHINE TYPE   AMI                     AGE
awsmachineclass.machine.sapcloud.io/shoot--xxx-worker-z1-a827a   m5.large       ami-x   24h
awsmachineclass.machine.sapcloud.io/shoot--xxx-worker-z1-ebc99   m5.large       ami-y   27h

I could delete the machine using CLI, but I think there needs to be a fix in the gardener/extension to handle such cases and update the old machineSet's credentials.

I think we need to dig deeper into the extensions code to fix these issues. I am unable to devote this time currently (due to migration of in-tree to OOT providers on prio), however I will take a look later if I can.

**Comment by rfranzke** [comment]
Hm, this incident looks different, to be honest. Have you checked whether the `Worker` object was still in the system?

> Bcoz the machine class had a deletion timestamp the credentials were not updated properly by the Gardener extension.

Why, i.e., how does it matter whether there is a deletion timestamp or not?

Usually, what we have seen as was that only `MachineClass`, `Secret`, and `ManagedResource` belonging to the worker extension remain in the system. Personally, I've never seen that any `Machine`, `MachineSet` or `MachineDeployment` is stuck (only if the credentials are invalid, but this scenario should be covered).

**Comment by prashanth26** [comment]
> Hm, this incident looks different, to be honest. Have you checked whether the Worker object was still in the system?

Yes, the worker object was still in the system in an errored status.

> Why, i.e., how does it matter whether there is a deletion timestamp or not?

Maybe I forgot to mention. But the shoot was in deletion. Hence I expected deletion timestamps to be set which it was. 

> Usually, what we have seen as was that only MachineClass, Secret, and ManagedResource belonging to the worker extension remain in the system. Personally, I've never seen that any Machine, MachineSet or MachineDeployment is stuck (only if the credentials are invalid, but this scenario should be covered).

Yes, it could be the case that this observed issue doesn't exactly fit this issue, we could probably open another issue if we observe this scenario again. And I think it could happen again as the same project has a similar shoot stuck in deletion with invalid credentials. And once this credential is updated, we might see it again. 

However, I did see this happening, where the `secrets of a machine class` was older than the one in the `cloud-credentials`. I had even reconciled the cluster a couple of times, just to be sure.

**Comment by rfranzke** [comment]
OK.
Probably https://github.com/gardener/gardener/blob/master/extensions/pkg/controller/worker/genericactuator/actuator_delete.go#L51-L61 is the section that we have to look at and improve.

**Comment by prashanth26** [comment]
I looked through the Gardener code a little for the deletion logic. 

Could this be the issue causing the orphan MachineClasses?

- https://github.com/gardener/gardener/blob/e512cd5a12edff2b4513fb6a97ead5f3f2c9909c/extensions/pkg/controller/worker/genericactuator/actuator_delete.go#L203
- https://github.com/gardener/gardener-extension-provider-azure/blob/master/pkg/controller/worker/machines.go#L49-L52

I am not sure how the secrets still exist though.

**Comment by ialidzhikov** [comment]
I was now checking similar issue. The azuremachineclass and the machineclass secret were orphan:

```
$ kubectl api-resources --verbs=list --namespaced -o name \
    6    | xargs -n 1 kubectl get --show-kind --ignore-not-found -n shoot--foo--bar


NAME                                      TYPE     DATA   AGE
secret/shoot--foo--bar-default-e73e8-z1   Opaque   5      19h

NAME                                                                     VM SIZE           AGE
azuremachineclass.machine.sapcloud.io/shoot--foo--bar-default-e73e8-z1   Standard_D4s_v3   19h
```

From the logs I see:

```

{"log":"{\"level\":\"info\",\"ts\":\"2020-08-25T12:23:56.920Z\",\"logger\":\"azure-worker-actuator.worker-actuator\",\"msg\":\"Deleting the machine-controller-manager\",\"worker\":\"shoot--foo--bar/bar\"}\n"}
{"log":"{\"level\":\"info\",\"ts\":\"2020-08-25T12:23:56.920Z\",\"logger\":\"azure-worker-actuator.worker-actuator\",\"msg\":\"Deleting all machines\",\"worker\":\"shoot--foo--bar/bar\"}\n"}
{"log":"{\"level\":\"info\",\"ts\":\"2020-08-25T12:23:53.722Z\",\"logger\":\"azure-worker-actuator.worker-actuator\",\"msg\":\"Deploying the machine classes\",\"worker\":\"shoot--foo--bar/bar\"}\n"}
{"log":"{\"level\":\"info\",\"ts\":\"2020-08-25T12:23:26.229Z\",\"logger\":\"azure-worker-actuator.worker-actuator\",\"msg\":\"Deploying the machine-controller-manager\",\"worker\":\"shoot--foo--bar/bar\"}\n"}
```

The azuremachineclass creation and deletion timstamps are:

```yaml
apiVersion: machine.sapcloud.io/v1alpha1
kind: AzureMachineClass
metadata:
  creationTimestamp: "2020-08-25T12:23:56Z"
  deletionTimestamp: "2020-08-25T12:33:29Z"
```

At `2020-08-25T12:23:53.722Z` the worker controller started the machine class deployment and at `2020-08-25T12:23:56.920Z` it started step `Deleting the machine-controller-manager` which is a step that occurs after the wait for cleanup of machineclasses and secrets. Note also the creationTimestamp of the machineclass - `2020-08-25T12:23:56Z`. It seems that the machineclass was not deleted after it was created and in the same second we were executing a task which is right after the machineclass cleanup. Looking at these timestamps I see that the machineclass was not deleted after it was created (it was deleted by the namespace deletion afterwards) but the deletion continued which can be possible if the list call for machineclasses returns empty list. Which makes me think that it can be caused by a cache issue (if cached client is used) or consistency issue on apiserver side.

**Comment by timebertt** [comment]
As discussed, let's add some more logs to the generic actuator, so that we can better analyze, when this issue comes up again.
/reopen
/assign
